flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -152,14 +153,35 @@ public String getLogicalScope(CharacterFilter filter) {
 	 * @return logical scope
 	 */
 	public String getLogicalScope(CharacterFilter filter, char delimiter) {
-		if (logicalScopeString == null) {
+		return getLogicalScope(filter, delimiter, -1);
+	}
+
+	/**
+	 * Returns the logical scope of this group, for example
+	 * {@code "taskmanager.job.task"}.
+	 *
+	 * @param filter character filter which is applied to the scope components
+	 * @param delimiter delimiter to use for concatenating scope components
+	 * @param reporterIndex index of the reporter
+	 * @return logical scope
+	 */
+	public String getLogicalScope(CharacterFilter filter, char delimiter, int reporterIndex) {
```

Comment:
This method could become package private and if I did not mess up, it could become something like:

```
String getLogicalScope(CharacterFilter filter, char delimiter, int reporterIndex) {
		if (logicalScopeStrings.length == 0 || (reporterIndex < 0 || reporterIndex >= logicalScopeStrings.length)) {
			return getChildLogicalScope(filter, delimiter);
		}

		if (logicalScopeStrings[reporterIndex] == null) {
			logicalScopeStrings[reporterIndex] = getChildLogicalScope(filter, delimiter);
		}
		return logicalScopeStrings[reporterIndex];
	}

	private String getChildLogicalScope(CharacterFilter filter, char delimiter) {
		final String groupName = getGroupName(filter);
		return parent == null
				? groupName
				: parent.getLogicalScope(filter, delimiter) + delimiter + groupName;
	}
```

I find this more readable and with less `return`s.